relayer/src/link/relay_path.rs

```diff
@@ -748,25 +780,28 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
     }
 
     /// Handles updating the client on the source chain
+    /// Returns the height at which the client update was processed
     fn update_client_src(
         &self,
         dst_chain_height: Height,
         tracking_id: &str,
-    ) -> Result<(), LinkError> {
-        if self
-            .src_chain()
-            .proven_client_consensus(self.src_client_id(), dst_chain_height, Height::zero())
-            .is_ok()
-        {
-            return Ok(());
+    ) -> Result<Height, LinkError> {
```

Comment:
The retry logic inside for loop is very difficult to understand and verified to work correctly. Can we split it into a recursive function that does the retry? Something like:

```rust
    /// Handles updating the client on the source chain
    /// Returns the height at which the client update was processed
    fn update_client_src(
        &self,
        dst_chain_height: Height,
        tracking_id: &str,
    ) -> Result<Height, LinkError> {
        if let Some(update_height) = Self::update_height(
            self.src_chain(),
            self.src_client_id().clone(),
            dst_chain_height,
        ) {
            Ok(update_height)
        } else {
            self.do_update_client_src(
                dst_chain_height,
                tracking_id,
                MAX_RETRIES,
            )
        }
    }

    /// Perform actual update_client_src with retries.
    ///
    /// Note that the retry is only performed in the case when there
    /// is a ChainError event. It would return error immediately if
    /// there are other errors returned from calls such as
    /// build_update_client_on_src.
    fn do_update_client_src(
        &self,
        dst_chain_height: Height,
        tracking_id: &str,
        retries_left: usize
    ) -> Result<Height, LinkError> {
        info!(
            "sending update_client to client hosted on source chain for height {} (retries left: {})",
            dst_chain_height,
            retries_left
        );

        let src_update = self.build_update_client_on_src(dst_chain_height)?;
        let tm = TrackedMsgs::new(src_update, tracking_id);

        let src_tx_events = self
            .src_chain()
            .send_messages_and_wait_commit(tm)
            .map_err(LinkError::relayer)?;

        info!("result: {}", PrettyEvents(&src_tx_events));

        for event in src_tx_events {
            match event {
                err_event @ IbcEvent::ChainError(_) => {
                    if retries_left == 0 {
                        return Err(LinkError::client(ForeignClientError::chain_error_event(
                            self.src_chain().id(),
                            err_event,
                        )))
                    } else {
                        return self.do_update_client_src(
                            dst_chain_height,
                            tracking_id,
                            retries_left - 1,
                        )
                    }
                }
                IbcEvent::UpdateClient(ev) => return Ok(ev.height()),
                _ => {}
            }
        }

        if retries_left == 0 {
            Err(LinkError::update_client_failed())
        } else {
            self.do_update_client_src(
                dst_chain_height,
                tracking_id,
                retries_left - 1,
            )
        }
    }
```

Reply:
Thanks for the suggestion, Soares! Done! 👌

Reply:
I wonder how we should handle the case where the `send_messages_and_wait_commit()` succeeds but the response contains neither an `UpdateClient` event nor a chain error. With this impl we'd be retrying and return a new kind of error if none of the retries work, but IIUC, we previously assumed that this was unlikely and called `src_err_ev.unwrap()` on the event.

Reply:
Seems like it is indeed possible that the returned events contain neither an `UpdateClient` event nor a chain error, which is when there is a `ClientMisbehaviour` event. So, I believe this implementation is doing the right thing by returning an `Err(LinkError::update_client_failed())`.

Reply:
https://github.com/informalsystems/ibc-rs/pull/1917/commits/9192eecaa2b5582d0277179b7dda984d27d83210